File: cpp-implementation/src/products/Books/Fiction.cpp
#include "Product.ih"

// Updated constructor implementation for Fiction books
Fiction::Fiction(int id, string name, double price, int volume,
                 string author, string publisher, string genre, string supplier)
    : Books(id, name, price, volume, author, publisher, supplier),
      genre(genre) {}

// Calculates the discount depending on the current month and the packaged quantity
double Fiction::calculateDiscount(int volume, int month) const {
    if (month == 6 && volume > 3) {
        return price * volume * 0.50;
    }
    return 0;
}

// Displays all the information
string Fiction::displayProductInfo() const {
    ostringstream info;
    info << "ID: " << id << "\nFiction Book: " << name << "\nSupplier: " << getSupplier()
         << "\nAuthor: " << author << "\nPublisher: " << publisher 
         << "\nGenre: " << genre << "\nPrice: €" << price 
         << "\nVolume: " << volume << " units\n\n";
    return info.str();
}

File: java-implementation/src/main/java/com/inventory/products/books/Fiction.java
package com.inventory.products.books;

// Represents a concrete subclass of Books
public class Fiction extends Books {
    private String genre;

    // Constructs an instance with specified details with base class attributes included
    public Fiction(int id, String name, double price, int volume, String author, String publisher, String genre, String supplier) {
        super(id, name, price, volume, author, publisher, supplier);
        this.genre = genre;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 6 && volume > 3) {
            return price * volume * 0.50;
        }
        return 0;
    }

    // Provides a string representation the details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nFiction Book: " + name + "\nAuthor: " + author
               + "\nPublisher: " + publisher + "\nGenre: " + genre
               + "\nPrice: €" + price + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}
